Add state transition panel setting to hide intermediate values

defunctzombie opened the pull request:
**User-Facing Changes**
<!-- will be used as a changelog entry -->

**Description**


<!-- link relevant GitHub issues -->
<!-- add `docs` label if this PR requires documentation updates -->
<!-- add relevant metric tracking for experimental / new features -->

amacneil commented:
- color state to end of recording

options:
- invert setting to "Show intermediate", off by default (i.e. performance by default)
- remove setting, show points based on heuristic (e.g. max half as many points as there are horizontal pixels)

defunctzombie commented:
This PR is ready for another round of product review and can get its first round of code review.

Regarding the first round feedback feedback; I've punted "color state to end of recording". I've decided that is better served by a separate follow-up discussion that does not block this change. Instead - I have updated this change to include the last input state value in the output state even if it is the same as before. This avoids "cutting off" the last bar prematurely since the last bar will always have the same state throughout.

Tooltips are always shown (even when between datums) as of: https://github.com/foxglove/studio/pull/7086

I have opted to avoid the heuristic approach in this PR to provide this incremental quality-of-life improvement and gather feedback. The heuristic approach (and doing more dataset truncation) will take more time to implement. The issue is more than hiding the points. Yes the points are slower to draw than not having the points - but even without drawing the points we have to avoid creating a large dataset for the worker architecture we currently have for this panel. That downsampling work will take more thinking.

snosenzo commented on packages/studio-base/src/panels/StateTransitions/messagesToDatasets.ts:
```diff
@@ -15,37 +15,23 @@ import positiveModulo from "./positiveModulo";
  */
-export default function messagesToDatasets(args: Args): ChartDatasets {
-  const { path, startTime, y, blocks } = args;
+export default function messagesToDatasets(args: Args): ChartDataset {
```
Can we update the function name and comment to have singular `Dataset` as well?

snosenzo commented on packages/studio-base/src/panels/StateTransitions/messagesToDatasets.ts:
```diff
@@ -82,18 +66,6 @@ export default function messagesToDatasets(args: Args): ChartDatasets {
 
-      const datasetIndex = indexer(value);
-
-      // Skip duplicates.
```
Why did we skip duplicates previously? and why don't we need to now?

defunctzombie replied:
I tried to think about why and could not figure it out nor come up with a reason for it. Maybe an optimization? @jtbandes do you recall?

jtbandes replied:
I do not know – seems to have been around since webviz days. Maybe it caused some chartjs glitch? Or maybe it was a premature optimization? The only way to know for sure is 🔪 and find out

snosenzo reviewed (approved):
code looks good to me. Might want another person to look it over who has more context on charts though

amacneil commented:
how much additional work is it to add the heuristic instead of toggle? it feels a bit weird adding a setting that we plan to delete, and it feels weird having a setting that makes the app explode.

defunctzombie commented:
Changed the setting to "Show points" and updated tooltip and behavior based on discussion that decided a more sensible behavior is to always hide all the points and have a setting to toggle points.
